app/services/step_update_service.rb

```diff
@@ -0,0 +1,51 @@
+class StepUpdateService
+  def initialize(step, order, params, session)
+    @step = step
+    @order = order
+    @params = params
+    @session = session
+  end
+
+  def update
+    public_send(@step.to_s) if respond_to? @step.to_s
+  end
+
+  def address
+    form = AddressForm.new(obj: @order)
+    form.billing_address.assign_attributes(address_params(:billing_form))
+    form.shipping_address.assign_attributes(address_params(:shipping_form))
+    form.save
+    form
+  end
+
+  def delivery
+    form = DeliveryForm.new(current_order: @order, delivery_id: @params[:order][:delivery])
+    form.save
+    form
+  end
+
+  def payment
+    form = CartForm.new(cart_params.to_h)
+    form.order = @order
+    form.save
+    form
+  end
+
+  def confirm
+    @order.waiting_processing!
+    @session[:steps_taken?] = false
+    @order
+  end
+
+  def complete; end
```

Comment:
I'm not sure it's a good idea to hack it this way

Reply:
It's service object for steps in checkout. The method 'complete' empty because there is nothing to update in 'complete' step.

Reply:
So, if we don't need to update something in complete step, why do we need to create this empty method?

Reply:
You have this check https://github.com/za4emnik/book_store/blob/dev/app/services/step_update_service.rb#L10
So... Your update method just return nil on complete state as well as an empty complete method, am I right? Just remove this empty method...

Reply:
Yup, update method should return nil.